galsim/table.py

```diff
@@ -280,3 +280,204 @@ def _LookupTable_hash(self):
 _galsim._LookupTable.__eq__ = _LookupTable_eq
 _galsim._LookupTable.__ne__ = lambda self, other: not self.__eq__(other)
 _galsim._LookupTable.__hash__ = _LookupTable_hash
+
+
+class LookupTable2D(object):
+    """
+    LookupTable2D represents a 2-dimensional lookup table to store function values that may be slow
+    to calculate, for which interpolating from a lookup table is sufficiently accurate.  A
+    LookupTable2D is also useful for evaluating periodic 2-d functions given samples from a single
+    period.
+
+    A LookupTable2D representing the function f(x, y) may be constructed from a list or array of `x`
+    values, a list or array of `y` values, and a 2D array of function evaluations at all
+    combinations of x and y values.  For instance:
+
+        >>> x = y = np.arange(5)
+        >>> z = x + y[:, np.newaxis]  # function is x + y, dimensions of z is (5, 5)
+        >>> tab2d = galsim.LookupTable2D(x, y, z)
+
+    To evaluate new function values with the lookup table, use the () operator:
+
+        >>> print tab2d(2.2, 3.3)
+        5.5
+
+    The () operator can also accept lists for the x and y arguments at which to evaluate the
+    LookupTable2D.  In this case, the table is evaluated for all combinations of x and y values and
+    returned as a 2D array with dimensions (len(y), len(x)).
+
+        >>> print tab2d([0, 1], [2, 3, 4])
+        [[ 2.  3.],
+         [ 3.  4.],
+         [ 4.  5.]]
+
+    Finally, if you want to just evaluate the LookupTable2D at a list of x and y points but without
+    evaluating all possible combinations of x and y values, then you can use the `scatter` keyword
+    with the () operator.
+
+        >>> print tab2d([1, 2], [3, 4], scatter=True)
```

Comment:
>So I wonder if it would make more sense to have this be the default, and maybe mesh=True do the other version (what you call Outer in the code).

Again, this is copied from scipy.  I guess I have a slight preference for keeping galsim.LookupTable2D as a drop-in replacement for scipy.interpolate.interp2d, but not a strong one if you think this is sufficiently counter-intuitive.

> Also, I know there is an optimization in the Outer code that saves some of the upperIndex calls, but do you know how much of a time savings that is?

Here's the comparison (including scipy too).  This is for interpolating a 200x200 grid from a 512x512 input grid.

SciPy interp2d
Function linear          took 0.00044 seconds.
GalSim LookupTable2D scatter
Function linear          took 0.00151 seconds.
GalSim LookupTable2D outer
Function linear          took 0.00095 seconds.


So outer is ~50% faster than scatter here.  Still a factor of 2 slower than scipy.  Should note that scipy requires its interpolation argument to be sorted, though, which I didn't want to require since it makes edge wrapping messier.